File: Weijia/213. House Robber II.py
class Solution(object):
    
    def helper(self, nums):
            if len(nums) ==1:
                return nums[0]
            elif len(nums) ==2:
                return max(nums[0],nums[1])
            else:
                m=[nums[0],max(nums[0],nums[1])]
                for i in range(2,len(nums)):
                    m=[m[-1],max(m[0]+nums[i],m[1])]
                return m[-1]
    def rob(self, num):
        if len(num)<=3:
            return max(num)
        t1=self.helper(num[1:])
        t2=self.helper(num[:-1])
        return max(t1,t2)

File: solution_zzj/213._House_Robber_II.cc
class Solution {
public:
    int rob(vector<int>& nums) {
        if (nums.size()==1)
            return nums[0];
        auto dp = [] (vector<int>& arr) {
            vector<int> ans = {arr[0], max(arr[0], arr[1])};
            for (int i=2;i<arr.size();i++) {
                ans.push_back(max(ans[i-2]+arr[i], ans[i-1]));
            }
            return ans[arr.size()-1];
        };
        int temp = nums[nums.size()-1];
        nums.pop_back();
        int a = dp(nums);
        cout<<a<<endl;
        nums.push_back(temp);
        nums.erase(nums.begin());
        return max(a, dp(nums));
    }
};
